### DISCUSSION TITLE
WebGPU compute rasterizer tutorial

### DISCUSSION THREAD

**Discussion by OmarShehata** [gpuweb/gpuweb#2409] [Show and tell]
Hey everyone!

I put together a tutorial + source code for building a simple rasterizer running on WebGPU compute shaders, with support for (basic) glTF models: 

https://github.com/OmarShehata/webgpu-compute-rasterizer

![image](https://user-images.githubusercontent.com/1711126/145739988-f5e1d4dd-c04f-4229-a7ad-f76d1ef5f800.png)

There is a full step-by-step guide here:

https://github.com/OmarShehata/webgpu-compute-rasterizer/blob/main/how-to-build-a-compute-rasterizer.md

I was inspired to create this after watching the [May 2021 WebGL/WebGPU talks](https://youtu.be/OIfqWD5NlNc?t=2441) where Markus Schütz showed a compute-based renderer that can render point clouds 10x faster compared to regular graphics pipeline. "Step 0" in the guide can be used as a starter kit for exploring compute based rendering techniques in WebGPU (it just sets up a simple compute shader that fills a storage buffer, and sends it to a fullscreen quad that draws it).

It's also a really fun exercise to have a software rasterizer running on the GPU like this. One effect that was really cool to explore was animating how triangles are filled in:

https://user-images.githubusercontent.com/1711126/145686026-c854ce90-e3ef-4750-a8dd-fc9842641772.mp4

Or visualizing in what order pixels are colored, especially while changing the workgroup size of the compute rasterizer:

https://user-images.githubusercontent.com/1711126/145685976-582dbf05-4e30-411a-9df9-a13e540cb3bc.mp4

I hope others find this helpful, especially if you're also interested in working with compute-based rendering techniques. Any feedback is welcome!

**Comment by Kangz** [comment]
That's really cool, thanks for sharing! I'm sure you are already aware of it but Unreal Engine 5's Nanite rendering component also uses compute shader rasterization for small (< 4x4) triangles as it can be done more efficiently than the HW rasterizer, at least on large GPUs.

**Reply by OmarShehata** [reply]
Thank you @Kangz! I had heard of that but had forgotten it. That sounds like another great concrete use case to add to my "motivation" section here, thank you for sharing that! I appreciate it.

**Comment by Rezmason** [comment]
This really is an excellent project and writeup, Omar!

**Reply by OmarShehata** [reply]
Thank you so much @Rezmason !
